contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,262 @@
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], _rewardsPercentagePPM, PPM_RESOLUTION);
+        }
+
+        // convert tokens to BNT
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            // skip token if no token balance
+            if (balances[i] == 0) {
+                continue;
+            }
+            // check if token is BNT - we don't need to swap in this case
+            if (token == _bnt) {
+                continue;
+            }
+
+            // get trade amount for token
+            uint256 tradeAmount = balances[i] - rewardAmounts[i];
+
+            // approve tokens for trading on Bancor Network V3
+            _setAllowance(token, tradeAmount);
+
+            uint256 val = token.isNative() ? tradeAmount : 0;
+
+            // swap token to BNT using Bancor Network V3
+            _bancorNetwork.tradeBySourceAmount{ value: val }(token, _bnt, tradeAmount, 1, block.timestamp, address(0));
+        }
+
+        // allocate rewards to caller and burn the rest
+        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    }
+
+    /**
+     * @dev allocates the rewards to caller and burns the rest
+     */
+    function _allocateRewards(address sender, Token[] calldata tokens, uint256[] memory rewardAmounts) private {
+        // transfer the rewards to caller
+        uint256 len = tokens.length;
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            uint256 rewardAmount = rewardAmounts[i];
+            if (rewardAmount == 0) {
+                continue;
+            }
+            if (token.isNative()) {
+                // using a regular transfer here could revert due to exceeding the 2300 gas limit
+                // which is why we're using call instead (via sendValue)
+                payable(sender).sendValue(rewardAmount);
+            } else {
+                token.safeTransfer(sender, rewardAmount);
+            }
+        }
+
+        // get the burn amount
+        uint256 burnAmount = _bnt.balanceOf(address(this));
```

Comment:
I believe that you can avoid this external call, and instead, accumulate the actual burn amount by summing up the values returned from the call to `_bancorNetwork.tradeBySourceAmount`.
In fact, it might even be the more correct thing to do, since the contract may be holding BNT for some different reason altogether (i.e., before this transaction is executed).

Reply:
Only case I see where this makes sense if the contract has BNT balance before, but the user doesn't send BNT as a token input, only other tokens - so the user in this case won't receive his 2% of the BNT burnt. Have to think a bit more about this case, it might make sense to accumulate the burn amount.

Reply:
Yup, that's the only case that I had in mind as well. And there's a good question of whether or we should include that additional balance as part of the process.
But regardless - my main point was to use information that you already have (the return-value of `tradeBySourceAmount`) in order to avoid the external function call (to `balanceOf` on the BNT contract).

Reply:
I get what you mean, but there are a lot of cases to consider here and there is a balance of keeping the code simpler and introducing new variables.
Consider that BNT gets passed as a token - we don't swap it - how are we going to get the BNT burn amount if we don't check the BNT balance ?
The return value of `tradeBySourceAmount` is not enough in this case, so we'll need to make another balanceOf call, and then sum that with the total accumulated burn amount. 
So if we don't want to make two calls to `balanceOf`, we're going to have to pass two variables to `_allocateRewards` (burn accumulator and bnt balance, which may be 0) just for this case.
I believe the current approach is simpler and any user who uses the vortex will want to maximize his rewards, thus if there is token balance of any token, he will send that token as well to the call.

Reply:
The return value of `tradeBySourceAmount` is of course not relevant for BNT, because we don't call that function in this case. We can just add the BNT amount to the total accumulator in this case. My point is to get rid of the `_bnt.balanceOf(this)` call at the end of the function, both for gas, as well as in order to really consider only the amount which was traded (excluding a potential previous balance).
Anyway, closing...